File: src/ProgenyImputerByOneParent.cpp
#include <cmath>
#include "../include/ProgenyImputerByOneParent.h"
#include "../include/common.h"

using namespace std;

ProgenyImputerByOneParent::ProgenyImputerByOneParent(
											const vector<VCFFamilyRecord *>& rs,
											const vector<vector<int>>& ref_hs,
											bool is_mat,
											const Map& map_, double w) :
										VCFHMM(rs, map_, w), ref_records(rs),
										ref_haps(ref_hs),
										is_mat_imputed(is_mat),
										Cc(calc_Cc(rs)), Cp(calc_Cp(rs)) { }

vector<double> ProgenyImputerByOneParent::calc_Cc(
							const vector<VCFFamilyRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

vector<double> ProgenyImputerByOneParent::calc_Cp(
							const vector<VCFFamilyRecord *>& rs) const {
	const double	K = 5.0;
	const size_t	M = rs.size();
	vector<double>	Cp(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cp[i] = Map::Kosambi(dist(rs[i], rs[i+1]) * K);
	}
	return Cp;
}

double ProgenyImputerByOneParent::emission_probability(size_t i, size_t j,
														int h,
														int parent_gt) const {
	const int	oc = ref_records[i]->unphased(j+2);
	const auto	alleles = parent_alleles(h, i);
	const int	mat_a = alleles.first;
	const int	pat_a = alleles.second;
	const int	phased_gt = Genotype::from_alleles(mat_a, pat_a) & 3;
	return E[phased_gt][oc];
}

vector<ProgenyImputerByOneParent::DP>
ProgenyImputerByOneParent::initialize_dp(size_t j) const {
	const size_t	L = NH() * 2;
	vector<DP>	dp(M(), DP(L, pair<double, int>(MIN_PROB, 0)));
	const VCFFamilyRecord	*record = ref_records[0];
	const int	parent_gt = is_mat_imputed ? record->mat_gt() :
												record->pat_gt();
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const double	E_all = emission_probability(0, j, h, parent_gt);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

void ProgenyImputerByOneParent::update_dp(size_t i, size_t j,
												vector<DP>& dp) const {
	const size_t	L = NH() * 2;
	const VCFFamilyRecord	*record = ref_records[i];
	const int	parent_gt = is_mat_imputed ? record->mat_gt() :
												record->pat_gt();
	
	for(int h = 0; h < (int)L; ++h) {
		const double	E = emission_probability(i, j, h, parent_gt);
		for(int prev_h = 0; prev_h < (int)L; ++prev_h) {
			const double	T = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (T + E);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void ProgenyImputerByOneParent::update_genotypes(size_t j,
													const vector<int>& hs) {
	for(size_t i = 0; i < this->M(); ++i) {
		const auto	alleles = parent_alleles(hs[i], i);
		const int	mat_a = alleles.first;
		const int	pat_a = alleles.second;
		const int	phased_gt = Genotype::from_alleles(mat_a, pat_a);
		ref_records[i]->set_geno(j+2, phased_gt);
	}
}

void ProgenyImputerByOneParent::impute(size_t j) {
	// DP
	vector<DP>	dp = initialize_dp(j);
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, j, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(j, hs);
}

File: python/ProgenyImputerByOneParent.py
from __future__ import annotations

# coding: utf-8
# ProgenyImputerByOneParent.py
# 片親がphasedでもう片親がunknownのときに後代をimputeする

from math import log
from typing import List, Tuple

from VCFFamily import VCFFamilyRecord
from VCFHMM import *
from Genotype import Genotype


#################### ProgenyImputerByOneParent ####################

MIN_PROB = -1e300

class ProgenyImputerByOneParent(VCFHMM[VCFFamilyRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[VCFFamilyRecord],
						ref_haps: list[list[int]],
						is_mat_imputed: bool, map_: Map, w: float) -> None:
		VCFHMM.__init__(self, records, map_)
		self.records: list[VCFFamilyRecord] = records
		self.ref_haps = ref_haps
		self.is_mat_imputed = is_mat_imputed
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(records, records[1:]) ]
		K = 5						# 親の遷移確率の子に対する倍率
		# 親は直接の後代でないので、乗り換え確率を高くする
		self.Cp = [ Map.Kosambi(self.dist(r1, r2) * K)
							for r1, r2 in zip(records, records[1:]) ]
	
	def num_progenies(self) -> int:
		return len(self.records[0].geno) - 2
	
	def NH(self) -> int:
		return len(self.ref_haps)
	
	def parent_alleles(self, h: int, i: int) -> tuple[int, int]:
		hc1 = h & 1
		hc2 = h >> 1
		a2 = self.ref_haps[hc2][i]
		if self.is_mat_imputed:
			a1 = (self.records[i].geno[0] >> hc1) & 1
			return (a1, a2)
		else:
			a1 = (self.records[i].geno[1] >> hc1) & 1
			return (a2, a1)
	
	# i: record index, j: progeny index
	def emission_probability(self, i: int, j: int, h: int,
											parent_gt: int) -> float:
		oc = self.records[i].unphased(j+2)
		mat_a, pat_a = self.parent_alleles(h, i)
		phased_gt = Genotype.from_alleles(mat_a, pat_a) & 3
		return self.E[phased_gt][oc]
	
	def transition_probability(self, i: int, prev_h: int, h: int) -> float:
		cc = self.Cc[i-1]	# 後代の遷移確率
		cp = self.Cp[i-1]
		hc1 = h & 1
		hc2 = h >> 1
		prev_hc1 = prev_h & 1
		prev_hc2 = prev_h >> 1
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if prev_hc1 != hc1 else 1.0 - cc) +
				log(cp if prev_hc2 != hc2 else 1.0 - cp))
	
	def initialize_dp(self, j: int, M: int) -> list[DP]:
		L = self.NH() * 2
		dp = [ [ (MIN_PROB, 0) ] * L for _ in range(M) ]
		record = self.records[0]
		parent_gt = record.mat_gt() if self.is_mat_imputed else record.pat_gt()
		for h in range(L):		# hidden state
			E = self.emission_probability(0, j, h, parent_gt)
			dp[0][h] = (E, h)
		return dp
	
	def update_dp(self, i: int, j: int, dp: list[DP]) -> None:
		L = self.NH() * 2
		record = self.records[i]
		parent_gt = record.mat_gt() if self.is_mat_imputed else record.pat_gt()
		
		for h in range(L):		# hidden state
			E = self.emission_probability(i, j, h, parent_gt)
			for prev_h in range(L):
				T = self.transition_probability(i, prev_h, h)
				prob = dp[i-1][prev_h][0] + (T + E)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, j: int, hs: list[int]) -> None:
		M = len(self.records)
		for i in range(M):
			record = self.records[i]
			mat_a, pat_a = self.parent_alleles(hs[i], i)
			record.geno[j+2] = Genotype.from_alleles(mat_a, pat_a)
	
	# j: index of progeny
	def impute(self, j: int) -> None:
		M = len(self.records)	# マーカー数
		
		# DP
		# 外側から、マーカー、ハプロタイプの状態
		dp = self.initialize_dp(j, M)
		for i in range(1, M):
			self.update_dp(i, j, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(j, hs)
